### ISSUE TITLE
Investigate using peribolos for knative github admininistration

### ISSUE THREAD

**Issue by srinivashegde86** [knative/test-infra#148]
https://github.com/kubernetes/test-infra/tree/master/prow/cmd/peribolos#peribolos-documentation

**Comment by srinivashegde86** [comment]
cc @jessiezcc , @BenTheElder

**Comment by adrcunha** [comment]
This should be taken to the TOC WG. It's mostly a matter of deciding what process we want to follow (i.e., if administration will be done through GitHub or through Prow). From the Perobolo's README I'm not really convinced that it currently adds value to Knative.

**Comment by srinivashegde86** [comment]
One use case, I thought of was, we need to ask mark or dewitt or other admin to add folks to knative org. If anyone interested can just create a PR and get added by default, it prevents a manual intervention by an admin. For eg. - https://knative.slack.com/archives/CA4DNJ9A4/p1537979117000100

We can send request to be added to knative-dev, but not for knative org.

**Comment by adrcunha** [comment]
Such process could be intentional (you don't have to be a member to contribute), so better taking this to the TOC WG as I suggested.

**Comment by evankanderson** [comment]
Added to TOC agenda for next week. Can someone write up a quick description or pros & cons?

https://docs.google.com/document/d/1hR5ijJQjz65QkLrgEhWjv3Q86tWVxYj_9xdhQ6Y5D8Q/edit#heading=h.5nc37ou2m7k4

**Comment by adrcunha** [comment]
/assign @srinivashegde86 

I volunteer Srinivas as a tribute. :D

**Comment by BenTheElder** [comment]
We're pretty early in using this in Lubernetes admittedly, but we pretty much have a join by PR process, see: https://github.com/kubernetes/org, EG https://github.com/kubernetes/org/pull/119

It's really nice having a public mechanism to do this and a public record of changes made to add people etc, similar to our test configs :-)
I'm hopeful we can expand this pattern (~GitOps) to more parts of the project. DNS management for the websites etc. is coming up soon.

**Comment by srinivashegde86** [comment]
## Pros:
* Publicly visible
* Auto-update from periobolos config to knative org.
* Easier to know who is part of a particular WG
* Dont need to wait on admins to add someone. Anyone joining can just send a PR to add themself to knative.
* Maintains running history
* Part of prow, so easy to integrate

## Cons:
* One more thing to maintain for knative-admins.
* May need frequent clean-up for members not part of the WG
* Not widely used.
* Debuggability. Maintained by prow team, but knative has its own prow cluster. So, knative/test-infra should be able to add and maintain this in our prow cluster.
* Monitoring. Do we need checks to make sure this tool works?

**Comment by BenTheElder** [comment]
I like those points. :+1: @fejta @cblecker how do we feel about monitoring for Kubernetes?

Is there a widely used non-manual alternative? If the tool fails for some reason admins can still add people manually, and there are checks to ensure we maintain a minimum admin set.

**Comment by fejta** [comment]
> May need frequent clean-up for members not part of the WG

How is this related to peribolos? If anything peribolos should make it easier to detect and remove random members. 

> Debuggability. Maintained by prow team, but knative has its own prow cluster. So, knative/test-infra should be able to add and maintain this in our prow cluster.

We just run this as a postsubmit prow job (and a periodic job to catch any erroneous human updates). Here's the entire job def: https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/test-infra/test-infra-trusted.yaml#L1-L23

> Monitoring. Do we need checks to make sure this tool works?

Monitoring doesn't make sense for something that runs 2m once a day. Just read the logs (ideally visible in testgrid). If you did something stupid fix it and rerun the tool. If it did something stupid send us a bug, use the new image we publish and rerun the tool. 

The tool automatically prevents you from doing something extremely stupid that you can't recover from (removing all the admins for example)

Here are the logs from our runs: https://k8s-testgrid.appspot.com/sig-testing-org#post-peribolos

**Comment by fejta** [comment]
Also what we did was submit the postsubmit job without the `--confirm` flag for ~week so we could see what it wanted to do without actually changing GitHub. 


Once we we're confident in it's behavior we added the `--confirm` flag to let it loose.

**Comment by srinivashegde86** [comment]
> How is this related to peribolos? If anything peribolos should make it easier to detect and remove random members.

It's not related to periobolos. But, someone will have to create a PR deleting folks that are no longer part of WG. It is an additional ask for the admins.

> Debuggability

Yes. It runs as another periodic job. But, someone will have to look at them, probably in testgrid. If it is failing, then someone from knative needs to isolate that failure to tool issue or cluster issue and triage accordingly.

> Monitoring

Again, someone has to read the logs. So, one additional task on some knative engineer. If we had some monitoring, we dont have to worry about doing this. Maybe periobolos will hardly fail and we dont need to automate this at all.

These are things we will have to do in addition to the current set of tasks admins/leads/engineers do. @fejta Added info about --confirm in Pros. There are a bunch of other Pros in the list too :)

The TOC WG can evaluate these in their WG meeting and decide if they want some automated process for knative Gitops or they prefer manual control without tool overhead.

**Comment by cblecker** [comment]
second everything @fejta just said. it was designed with lots of safeguards (such as manually specifying admins via a flag that it will fatal if the config is missing them, and delta measurement to reject mass removal of users).

the biggest advantage is the one kubernetes is just now implementing: delegation of administrative privs. an unprivileged user can do things like suggesting new members, or removing inactive members via a PR, and these can be merged as part of a git workflow (we do it with OWNERS files to delegate permissions)

obviously, you'll want to choose for yourself if this actually adds value to your community, but feel free to ask any questions you might have.

**Comment by fejta** [comment]
> There are a bunch of other Pros in the list too :) The TOC WG can evaluate these in their WG meeting 

Sounds good. Just trying to clarify some things. Can be quiet too if you prefer.  

> Dont need to wait on admins to add someone. Anyone joining can just send a PR to add themself to knative.

Hopefully this still requires random internet user to wait for admins (or some other trusted set of org maintainers) to approve the PR.

> someone will have to create a PR deleting folks that are no longer part of WG. It is an additional ask for the admins.

I still don't understand this? Either admins have to click a bunch of buttons on GitHub to remove the users or else they need to merge a PR. The theory is the PR method is more familiar and provides better auditing. 

> If we had some monitoring, we dont have to worry about doing this.

You can configure testgrid to send an email when jobs fail. Alerting still requires someone to spend the time respond to the alert/email.

**Comment by srinivashegde86** [comment]
## Meeting notes:
* Investigate and create a prototype.
* Not super high priority. Nice to have for now.
* Knative admins to follow-up for public/private repos.

**Comment by jessiezcc** [comment]
Closing as no recent asks and not a priority.

**Comment by adrcunha** [comment]
Reopening as this came up again in the TOC meeting today.

/reopen

**Comment by evankanderson** [comment]
Would it make sense to put these in a sub-directory of https://github.com/knative/community ?

**Comment by coryrc** [comment]
Since he's actually doing it:
/assign @evankanderson 

/unassign steuhs
